File: Problem2.cs
//Time complexity O(1) as it is array and direct access to memory
//Space Complexity = O(n) as we have to allocate the memory for 10^6
public class MyHashSet
    {

        public bool[,] storage;
        public int buckets;
        public int bucketItems;
        /** Initialize your data structure here. */
        public MyHashSet()
        {
            buckets = 1000;
            bucketItems = 1001;
            storage = new bool[buckets, bucketItems];
        }
        private int GetBucket(int key)
        {
            return key % buckets;
        }
        private int GetBucketItems(int key)
        {
            return key / bucketItems;
        }
        public void Add(int key)
        {
            int bucket = GetBucket(key);
            int bucketItem = GetBucketItems(key);
            if (storage[bucket, bucketItem] == false)
            {
               storage[bucket, bucketItem] = true;
            }
        }

        public void Remove(int key)
        {
            int bucket = GetBucket(key);
            int bucketItem = GetBucketItems(key);
            if (storage[bucket, bucketItem] == false) return;
            storage[bucket, bucketItem] = false;
        }

        /** Returns true if this set contains the specified element */
        public bool Contains(int key)
        {
            int bucket = GetBucket(key);
            int bucketItem = GetBucketItems(key);
            if (storage[bucket, bucketItem] == false) return false;
            return storage[bucket, bucketItem];
        }
    }

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet obj = new MyHashSet();
 * obj.Add(key);
 * obj.Remove(key);
 * bool param_3 = obj.Contains(key);
 */

File: problem2.cpp
// Time Complexity : O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes

//using arrays

class MyHashMap {
public:
    int buckets = 1000;
    int bucketItems = 1000;
    vector<vector<int>>storage = vector<vector<int>>(1000,vector<int>());
    MyHashMap() {
        
    }
    int getbucket(int key){
        return key%buckets;
    }
    int getbucketItem(int key){
        return key/bucketItems;
    }
    
    void put(int key, int value) {
        int bucket = getbucket(key);
        if(storage[bucket].empty()){
            storage[bucket] = vector<int>(1000,-1);
        }
        int bucketItem = getbucketItem(key);
        if(bucketItem == buckets){
            storage[bucket].push_back(value);
        }
        else
        storage[bucket][bucketItem]= value;        
        
    }
    
    int get(int key) {
        int bucket = getbucket(key);
        if(storage[bucket].empty()){
            return -1;
        }
        int bucketItem = getbucketItem(key);
        return storage[bucket][bucketItem];
        
    }
    
    void remove(int key) {
        int bucket = getbucket(key);
        if(storage[bucket].empty()){
            return;
        }
        int bucketItem = getbucketItem(key);
        storage[bucket][bucketItem]= -1;           
    }
};

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key,value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */